packages/backend-common/src/database/postgres.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { PgConnectionConfig } from 'knex';
+import { Config } from '@backstage/config';
+
+/**
+ * Builds a knex postgres connection config
+ *
+ * @param config The database config
+ * @param database The optional database name
+ */
+export function buildPgConnectionConfig(
+  config: Config,
+  database?: string,
+): PgConnectionConfig {
+  const connection = config.get('connection');
+
+  if (database) {
+    return replaceDatabase(connection, database);
+  }
+
+  return connection as PgConnectionConfig;
+}
+
+/**
+ * Parses a standard postgres uri into a knex connection config
+ *
+ * @param connectionString The postgres uri connection string
+ */
+export function parsePgConnectionString(
```

Comment:
why? can't we just pass it along?

https://knexjs.org/
search
`connection: process.env.PG_CONNECTION_STRING`

Reply:
@timja - If you follow a little closer in the `buildPgConnectionString` + `replaceDatabase` you'll see it actually works this way, unless you want to swap out the database name for each plugin (which requires parsing the `uri`).

For instance, this would work for a single database (e.g. `postgresql://postgres@localhost/backstage`)

```yaml
backend:
  database:
    connection:
      $secret:
        env: PG_CONNECTION_STRING
```

But this won't work without modifying the the other plugin specific databases which would need to be `postgresql://postgres@localhost/backstage_plugin_catalog`. This is an attempt to make it possible to re-use an single connection string with a convention for separate plugins databases. Very open to alternatives, but the hope was to avoid having to create a bunch of separate database configs for each plugin db.

Reply:
my biggest concern here is that it's quite common for ssl to be required for DBs, and if you parse it out you'll have to handle parsing the relevant options too

>  If you follow a little closer in the buildPgConnectionString + replaceDatabase you'll see it actually works this way, unless you want to swap out the database name for each plugin (which requires parsing the uri).

I don't follow, it doesn't look like it's passing the options along, it appears to pick out specific options and then replaces the database config value

Reply:
Agreed about the SSL part. In my `todo` I wanted to add support for this as well. It's really only an issue when using a connection string uri.

If you see this [line](https://github.com/spotify/backstage/pull/1915/files#diff-35f129e25b2fd534388750cc73f3e702R32) is the only path that does not use a pass through. Then in [replaceDatabase](https://github.com/spotify/backstage/pull/1915/files#diff-35f129e25b2fd534388750cc73f3e702R66) is the only time it actually parses the string down.

Reply:
@timja - I refactored this a bit and made it so you can override any arbitrary options. Does this seem like a reasonable approach to the api?

```ts
const database = createDatabase(config.getConfig('backend.database'), {
  database: `backstage_plugin_${plugin}`,
  ssl: {
    rejectUnauthorized: true,
    ca: fs.readFileSync(config.getString('backend.pathToRootCert')),
  }
})
```

Reply:
Yes it looks reasonable to me? But just to clarify I’m reading it right,

1. If your config is an object you can pass arbitrary config and it will just get passed along
2. If you’re using a connection string only the fields in the object on line 89 will be extracted, I.e ssl won’t work with connection string
3. Where it that file read going to be done for the ca certificate?

Reply:
After some further digging, I found `pg-connection-string` which is actually what `node-postgres` uses under the covers. Might consider swapping this custom code out to use that library as a peer dependency (only when you want to override a connection string property).